### PR TITLE
Adding IPv4 specific loading code

### PR DESCRIPTION
**Pull request by asergeyev**
Follow up on #166 ... I don't know some internals so please feel free to disagree with this way of plugging it in.

### CODE REVIEW

**Comment by asergeyev** [comment]
found another place, fixing

**Comment by asergeyev** [comment]
I think now it should be doing what's intended in #166

File: zscan_rr.go [hunk 1]
```diff
@@ -136,7 +136,7 @@ func endingToTxtSlice(c chan lex, errstr, f string) ([]string, *ParseError, stri
 	return s, nil, l.comment
 }
 
-func setA(h RR_Header, c chan lex, o, f string) (RR, *ParseError, string) {
+func setA(h RR_Header, c chan lex, o, f string) (_ RR, e *ParseError, _ string) {
```
Comment (miekg) [inline zscan_rr.go]: Could you not use named params here?

File: zscan.go [hunk 2]
```diff
@@ -962,3 +963,44 @@ func stringToNodeID(l lex) (uint64, *ParseError) {
...
+func stringToIPv4(l lex) (net.IP, *ParseError) {
+	var (
+		ip  uint64
+		oct uint32
+		b   byte
+		num byte
+	)
+
+	// not checking len upfront, error will be obvious in soon anyway
+	for _, b = range []byte(l.token) {
```
Comment (miekg) [inline zscan.go]: why does this need to be converted to []byte? We could use the string as-is, no? And convert a single value to a byte?

File: zscan.go [hunk 3]
```diff
@@ -962,3 +963,44 @@ func stringToNodeID(l lex) (uint64, *ParseError) {
...
+		switch {
+		case b == '.':
+			if num++; num > 3 {
```
Comment (miekg) [inline zscan.go]: why not use the counter you get back from `range`?

File: zscan.go [hunk 4]
```diff
@@ -962,3 +963,44 @@ func stringToNodeID(l lex) (uint64, *ParseError) {
...
+		case b >= '0' && b <= '9':
+			oct = oct*10 + uint32(b-'0')
+			if oct > 255 {
+				return nil, &ParseError{l.token, "bad IP address", l}
+			}
+		default:
+			return nil, &ParseError{l.token, "bad IP address", l}
+		}
+	}
+	if num < 3 {
```
Comment (miekg) [inline zscan.go]: ah you need `num` here. OK.

File: zscan_rr.go (hunk 1 above)
Comment (asergeyev) [inline zscan_rr.go]: That's lazy.. I should just reverse that and add var e *ParseError few lines lower. 
Named params just not going to be used, rr has to have type A and comment will also stay unused.

File: zscan.go (hunk 3 above)
Comment (asergeyev) [inline zscan.go]: that's for full octets, not for individual characters

**Comment by asergeyev** [comment]
Thanks, I'll get fixes for []byte() after mem test.

**Comment by asergeyev** [comment]
Well... Fixes for []byte and other code review notes separated from memory test. I decided to put things in same branch and test with hack that swaps parser to old function...

Looks like many small slices end up allocating with padding to fit more data later....  I see buckets size 8 and 16 go up when test is ran

**Comment by asergeyev** [comment]
Hard to call memprofiler accurate, but when I crank LENGTH from 0 to 1000 it seem there is some extractable numbers 

HeapObjects growth always smaller on new ip parser
HeapInUse always bigger on new ip parser
TotalAlloc bigger on old parser but closer to new one if tests swapped in order

On million A recs to parse new one wins on all numbers all the time regardless of where it placed... But IMO those wins are relatively small.

**Comment by asergeyev** [comment]
After changing this code to be a benchmark instead:

		$ go test -v -run=nothing -bench=BenchmarkZoneParseWith
		PASS
		BenchmarkZoneParseWithStringToIPv4	   10000	    127196 ns/op
		--- BENCH: BenchmarkZoneParseWithStringToIPv4
			mem_test.go:57: Average alloc growth per iteration: 111520 (1 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 98664 (2 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 91289 (5 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 87604 (20 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86621 (100 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86457 (300 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86400 (1000 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86384 (3000 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86378 (10000 parsing rounds)
		BenchmarkZoneParseWithNetParseIP	   10000	    124895 ns/op
		--- BENCH: BenchmarkZoneParseWithNetParseIP
			mem_test.go:57: Average alloc growth per iteration: 111488 (1 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 98928 (2 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 91398 (5 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 87631 (20 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86627 (100 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86459 (300 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86401 (1000 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86384 (3000 parsing rounds)
			mem_test.go:57: Average alloc growth per iteration: 86378 (10000 parsing rounds)
		ok  	github.com/miekg/dns	9.897s

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [dns] Adding IPv4 specific load..." ]
>Hard to call memprofiler accurate, but when I crank LENGTH from 0 to 1000 it seem there is some extractable numbers
>
>HeapObjects growth always smaller on new ip parser
>HeapInUse always bigger on new ip parser
>TotalAlloc bigger on old parser but closer to new one if tests swapped in order
>
>On million A recs to parse new one wins on all numbers all the time regardless of where it placed... But IMO those wins are relatively small.

So, not worth merging in your opinion?

We could still do a net.Parse().To4() copy out the bytes and destroy (or let gc
take care of the unused value).

**Comment by asergeyev** [comment]
I think it's worth merging but I know that fact that I wrote that code plays a role in making my opinion...
* functionally we do not change much, code is still readeable
* we do not do anything that makes code slower
* overall memory consumption may go down (I'd speculate that there is possibility that in certain conditions go would choose 4-byte objects to store our 4-byte slices, we just do not see this happen in tests)

Copying bytes would waste another 8 or 16 bytes (whichever bucket go would choose for new make()), it's unlikely to have just 4 byte impact. Also many memmove would slow thigns down. Also, it seem to me that even calling To4 when you don't need to is worse than just keeping 16 bytes, new slice eats new 12 bytes off stack to save cap/len values (reason why in last version of test I removed that, also matches current state of the library)

Against the merge is the fact that **there is nothing wrong I can see with current solution**. Most users expect there net.IP in structs, we could not get to revolutionize too much.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [dns] Adding IPv4 specific load..." ]
>I think it's worth merging but I know that fact that I wrote that code plays a role in making my opinion...
>* functionally we do not change much, code is still readeable
>* we do not do anything that makes code slower
>* overall memory consumption may go down (I'd speculate that there is possibility that in certain conditions go would choose 4-byte objects to store our 4-byte slices, we just do not see this happen in tests)
>
>Copying bytes would waste another 8 or 16 bytes (whichever bucket go would choose for new make()), it's unlikely to have just 4 byte impact. Also many memmove would slow thigns down. Also, it seem to me that even calling To4 when you don't need to is worse than just keeping 16 bytes, new slice eats new 12 bytes off stack to save cap/len values (reason why in last version of test I removed that, also matches current state of the library)
>
>Against the merge is the fact that **there is nothing wrong I can see with current solution**. Most users expect there net.IP in structs, we could not get to revolutionize too much.

That last point is intersting... I'm leaning against closing this issue and
*not* merging this PR...

**Comment by asergeyev** [comment]
Sure :) I actually agree.

**Comment by asergeyev** [comment]
You might want to close #166.

**Comment by miekg** [comment]
Thanks!

**Comment by miekg** [comment]
done

**Pull request closed without merging**